src/renderer/renderer.rs

```diff
@@ -0,0 +1,314 @@
+use super::fonts::atlas::Atlas;
+use super::pipeline::{Background, BackgroundFragment, Camera, GlyphFragment, Glyphs};
+use bytemuck::{cast_slice, Pod, Zeroable};
+use csscolorparser::Color;
+use euclid::default::Size2D;
+use hotwatch::Hotwatch;
+use pollster::FutureExt as _;
+use std::mem::size_of;
+use std::ops::Range;
+use std::sync::{
+    atomic::{AtomicBool, Ordering},
+    Arc,
+};
+use wgpu::{
+    util::{BufferInitDescriptor, DeviceExt},
+    *,
+};
+use winit::window::Window;
+
+#[repr(C)]
+#[derive(Copy, Clone, Debug, Pod, Zeroable)]
+pub struct QuadVertex {
+    position: [f32; 2],
+}
+
+impl QuadVertex {
+    const ATTRIBS: [VertexAttribute; 1] = vertex_attr_array![0 => Float32x2];
+
+    pub fn desc<'a>() -> VertexBufferLayout<'a> {
+        VertexBufferLayout {
+            array_stride: size_of::<Self>() as BufferAddress,
+            step_mode: VertexStepMode::Vertex,
+            attributes: &Self::ATTRIBS,
+        }
+    }
+}
+
+const QUAD_VERTICES: &[QuadVertex] = &[
+    QuadVertex {
+        position: [0.0, 0.0],
+    },
+    QuadVertex {
+        position: [1.0, 0.0],
+    },
+    QuadVertex {
+        position: [1.0, 1.0],
+    },
+    QuadVertex {
+        position: [0.0, 1.0],
+    },
+];
+
+const QUAD_INDICES: &[u16] = &[0, 3, 1, 3, 2, 1];
+
+pub struct WGpuRenderer {
+    surface: Surface,
+    pub device: Device,
+    pub queue: Queue,
+    surface_config: SurfaceConfiguration,
+    size: winit::dpi::PhysicalSize<u32>,
+    quad_vertex_buffer: Buffer,
+    quad_index_buffer: Buffer,
+    camera: Camera,
+    background: Background,
+    glyphs: Glyphs,
+    validation_errors_shown: Arc<AtomicBool>,
+    hotwatcher: Hotwatch,
+}
+
+pub struct MainRenderPass<'a> {
+    render_pass: RenderPass<'a>,
+    background: &'a Background,
+    glyphs: &'a Glyphs,
+    queue: &'a Queue,
+    device: &'a Device,
+    quad_vertex_buffer: &'a Buffer,
+    quad_index_buffer: &'a Buffer,
+}
+
+impl WGpuRenderer {
+    pub fn new(window: &Window) -> Self {
+        async {
+            let size = window.inner_size();
+
+            // The instance is a handle to our GPU
+            // Backends::all => Vulkan + Metal + DX12 + Browser WebGPU
+            let instance = Instance::new(Backends::DX12);
```

Comment:
Use `Backends::ALL`?

Reply:
Yes, I forced DX12 for better integration with tools like Pix for Windows, because Vulkan is the default on Windows. 

I think we might want to use DX12 even in the final version on Windows, but with a command line option to switch to Vulkan, so some platform specific code is probably needed here.

Reply:
Wgpu DX12 currently uses way too much RAM for an app like neovide (I'll try and get a PR up soon that should be a sizeable reduction, but it'll still be pretty high). If you're just testing with it, you can set the backend with `wgpu::util::backend_bits_from_env` ([example](https://github.com/gfx-rs/wgpu/blob/2571af9557eeba59f00f69649ae6bcd9d5a62392/wgpu/examples/framework.rs#L160)) and use the `WGPU_BACKEND` environment variable (`$env:WGPU_BACKEND="dx12"`, or vulkan/gl/etc in the terminal to temporarily change the env variable for that terminal session and anything launched from that terminal).